### DISCUSSION TITLE
migration guides to slack_cleaner2

### DISCUSSION THREAD

**Discussion by sgratzl** [sgratzl/slack-cleaner#97] [Q&A]
Slack Cleaner v2 at https://github.com/sgratzl/slack_cleaner2 is an improved version of this slack cleaner script. It is easier to maintain and more powerful. However, instead of a single script to execute it is a Python library module. Thus, it requires to code in Python. 

This issue is a collection of common examples in `slack-cleaner ` and their counterpart in `slack_cleaner2`

In each example `<TOKEN>` needs to be replaced by the corresponding Slack API token

**Comment by sgratzl** [comment]
## Show Information about the Slack Workspace

```sh
slack-cleaner --token <TOKEN> --info
```

will become

`info.py`
```py
from slack_cleaner2 import SlackCleaner
s = SlackCleaner("<TOKEN>")
print('users', s.users)
print('public channels', s.channels)
print('private channels', s.groups)
print('instant messages', s.ims)
print('multi user direct messages', s.mpim)
```
```sh
python info.py
```

**Comment by sgratzl** [comment]
## Delete all messages from the general channel

```sh
slack-cleaner --token <TOKEN> --rate 1 --message --channel general --user "*"
```
will become

clean_general.py
```python
from slack_cleaner2 import SlackCleaner
s = SlackCleaner("<TOKEN>", sleep_for=1)
for msg in s.c.general.msgs(with_replies=True):
  msg.delete()
```
```sh
python clean_general.py
```

**Comment by sgratzl** [comment]
## Delete all messages from bots (especially flooding CI updates)
```sh
slack-cleaner --token <TOKEN> --rate 1 --message --channel auto-build --bot
```
will become

clean_bot.py
```python
from slack_cleaner2 import SlackCleaner
s = SlackCleaner("<TOKEN>", sleep_for=1)
for msg in s.c['auto-build'].msgs():
  if msg.bot:
    msg.delete()
```
```sh
python clean_bot.py
```

**Comment by sgratzl** [comment]
## Delete all messages older than 2015/09/19

```sh
slack-cleaner --token <TOKEN> --rate 1 --message --channel general --user "*" --before 20150919
```
will become

clean_general_before2015.py
```python
from slack_cleaner2 import SlackCleaner
s = SlackCleaner("<TOKEN>", sleep_for=1)
for msg in s.c.general.msgs(before="20150919", with_replies=True):
  msg.delete()
```
```sh
python clean_general_before2015.py
```

**Comment by jpwynn** [comment]
what if user is "sam" instead of "*" which seems to be the default? IS there a way to filter for that?

**Comment by sgratzl** [comment]
## Delete all messages in channel general by user named sam but keep pinned messages

```sh
slack-cleaner --token <TOKEN> --rate 1 --message --channel general --user "sam" --keeppinned
```
will become

clean_general_sam.py
```python
from slack_cleaner2 import SlackCleaner
s = SlackCleaner("<TOKEN>", sleep_for=1)
for msg in s.c.general.msgs(with_replies=True):
  if msg.user.name == 'sam' and not msg.pinned_to:
    msg.delete()
```
```sh
python clean_general_sam.py
```

alternative using predicate syntax (untested)

clean_general_sam_predicate.py
```python
from slack_cleaner2 import SlackCleaner, is_not_pinned, match_user
s = SlackCleaner("<TOKEN>", sleep_for=1)
for msg in filter(is_not_pinned() and match_user('sam'), s.c.general.msgs(with_replies=True)):
  msg.delete()
```

**Reply by randpop** [reply]
Maybe i understand something wrong, but i always get the error message
'AttributeError: 'NoneType' object has no attribute 'name'
when trying to the first snippet. Double checked all lines and tried different user names.
Any idea why this is happening?
BTW: Awesome work!

**Reply by sgratzl** [reply]
it could be that the `.user` couldn't be resolved and thus it is not defined. You could. just use `if msg.user and msg.user.name == 'sam' and not msg.pinned_to:` to check for it

**Reply by randpop** [reply]
Worked like charm. Thank you so much!

**Comment by Freelensia** [comment]
Hi @sgratzl is there a way to clean msg in multiple channels instead of repeating the entire script?
And btw I stumbled on this issue only by chance! These sample commands should be in the guide of the slack-cleaner2 github page.

**Comment by sgratzl** [comment]
## Delete all messages in all channels

```sh
slack-cleaner --token <TOKEN> --rate 1 --message --channel ".*" --regex --user "*"
```

will become

clean_all.py
```python
from slack_cleaner2 import SlackCleaner
s = SlackCleaner("<TOKEN>", sleep_for=1)

for channel in s.conversations:
  for msg in channel.msgs(with_replies=True):
    msg.delete()
```
```sh
python clean_all.py
```

**Comment by orappold** [comment]
Hey sgratzl, 
would you know a smart way to only delete message with no replies, e.g. no Thread has been started? The challenge I guess is to completely exclude a thread (parent + child entries) from the deletion process. I was not able to come up with a solution for that. Would be happy to hear your opinion!

**Comment by sgratzl** [comment]
> Hey sgratzl,
> would you know a smart way to only delete message with no replies, e.g. no Thread has been started? The challenge I guess is to completely exclude a thread (parent + child entries) from the deletion process. I was not able to come up with a solution for that. Would be happy to hear your opinion!

```py
from slack_cleaner2 import SlackCleaner
s = SlackCleaner("<TOKEN>", sleep_for=1)

for channel in s.conversations:
  for msg in channel.msgs():
    if not next(msg.replies(), None): # replies are empty
       msg.delete()
```

or using some internal knowledge what makes message starting a thread start
```py
from slack_cleaner2 import SlackCleaner
s = SlackCleaner("<TOKEN>", sleep_for=1)

for channel in s.conversations:
  for msg in channel.msgs():
    if 'thread_ts' not in msg.json: #reply thread_ts info is not set
       msg.delete()

**Comment by leonard84** [comment]
Why not offer both, i.e., users that need more control code in python users that just want to delete message from a channel use the script args and this is internally mapped to what you've written here?

**Comment by rogerdahl** [comment]
Delete all files:
```python3

#!/usr/bin/env python

from slack_cleaner2 import SlackCleaner

s = SlackCleaner('<TOKEN (xoxp-...)>')

for file in s.files():
    print(f'Deleting: {file}')
    file.delete()
```

**Comment by rogerdahl** [comment]
> Why not offer both, i.e., users that need more control code in python users that just want to delete message from a channel use the script args and this is internally mapped to what you've written here?

Agreed - adding a thin command line client on top of the library, that handles common cases seems like it would take just about the same amount of time as listing the code for the common cases on this page.

At the same time, I really appreciate what @sgratzl has already provided for us here.

**Comment by ghost** [comment]
Ok

**Comment by aeron7** [comment]
> Hi @sgratzl is there a way to clean msg in multiple channels instead of repeating the entire script?
> And btw I stumbled on this issue only by chance! These sample commands should be in the guide of the slack-cleaner2 github page.

```
channelvars = ['daytraders','basket']

for channelvar in channelvars:
  for msg in s.c[channelvar].msgs(with_replies=True):      
    msg.delete()
```

**Comment by nasalaika** [comment]
I'm wondering if someone would be kind enough to post an example of deleting a whole thread of direct (instant) messages with another specific user, say Leroy? Thanks in advance - I'm only half able to understand the available examples and my python skills are not what I'd hope they'd be at the moment...

Below is what I've been attempting, but I think there's loads wrong....

```
from slack_cleaner2 import SlackCleaner
s = SlackCleaner("TOKEN_HASH_HERE", sleep_for=1)

for msg in s.ims.leroy.msgs(with_replies=True):
  msg.delete()
```

**Comment by k4rtik** [comment]
> ```python
> from slack_cleaner2 import SlackCleaner
> s = SlackCleaner("<TOKEN>", sleep_for=1)
> for msg in s.c['auto-build'].msgs():
>   if msg.bot:
>     msg.delete()
> ```

Is there a way to specify `botname` here? the `msg` object doesn't seem to have that attribute.

> AttributeError: 'SlackMessage' object has no attribute 'botname'

**Reply by steven-geo** [reply]
msg.json['username'] should get you the botname

**Comment by jameshibbard** [comment]
> I'm wondering if someone would be kind enough to post an example of deleting a whole thread of direct (instant) messages with another specific user, say Leroy?

@nasalaika: you can do it like this. 

```python
from slack_cleaner2 import SlackCleaner
s = SlackCleaner(<TOKEN>, sleep_for=1)

for c in s.ims:
  if c.name == 'leroy':
    for msg in c.msgs():
      if msg.user == s.myself:
        msg.delete()
```

Bear in mind that the Slack API doesn't permit the deletion of direct messages that are not written by you, so Leroy would have to delete his half of the conversation.

**Comment by sgratzl** [comment]
> 'NoneType' object has no attribute 'msgs'  why see this?
> ```

seems like it couldn't find the channel named `auto-build`

**Comment by devaeterne** [comment]
> > Hey sgratzl,
> > would you know a smart way to only delete message with no replies, e.g. no Thread has been started? The challenge I guess is to completely exclude a thread (parent + child entries) from the deletion process. I was not able to come up with a solution for that. Would be happy to hear your opinion!
> 
> ```python
> from slack_cleaner2 import SlackCleaner
> s = SlackCleaner("<TOKEN>", sleep_for=1)
> 
> for channel in s.conversations:
>   for msg in channel.msgs():
>     if not next(msg.replies(), None): # replies are empty
>        msg.delete()
> ```
> 
> or using some internal knowledge what makes message starting a thread start
> 
> ```python
> from slack_cleaner2 import SlackCleaner
> s = SlackCleaner("<TOKEN>", sleep_for=1)
> 
> for channel in s.conversations:
>   for msg in channel.msgs():
>     if 'thread_ts' not in msg.json: #reply thread_ts info is not set
>        msg.delete()
> ```


cannot delete entry: xxxxx:1624088712.0007 (xxxxxxxx (xxxxxxxxxxxxx) SERHAT BARLAS) cant_delete_message

When I want to delete all messages, I cannot delete the messages of the error user

**Comment by pedro380085** [comment]
delete all messages from a specific thread, you just need to pass the url as an argument.


```
from slack_cleaner2 import SlackCleaner
from pprint import pprint
import sys

thread_identifier = sys.argv[1]

# get the last part of the string
thread_identifier = thread_identifier.split("/")
thread_identifier = thread_identifier[-1]

# remove the first p
if thread_identifier[0] == "p":
    thread_identifier = thread_identifier[1:]

# remove last 2 zeros
thread_identifier = thread_identifier[-2]

# add : to the middle of the string
thread_index = len(thread_identifier) - 4;
thread_identifier = thread_identifier[:thread_index] + "." + thread_identifier[thread_index:]

print("thread = ", thread_identifier)
# sys.exit()

s = SlackCleaner("TOKEN")

for msg in s.c.general.msgs(with_replies=True):
  if str(msg.thread_ts) == str(thread_identifier):
    msg.delete()
```

**Comment by agardony** [comment]
Hi sgratzl,

I'm trying to download files within a specific IM. When I run the below code I can only download the same 10 files at a time. If if I delete those files I can get 10 new ones but it seems I'm locked into a limit of 10 files at a time. Is file download count limited to 10 at a time?  If so, can this be altered to support downloading a larger numbers of files at a time?

```
from slack_cleaner2 import SlackCleaner
s = SlackCleaner("<TOKEN>", sleep_for=1)

for c in s.ims:
	if c.name == 'user123':
		for file in c.files():
			print(f'Downloading: {file}')
			file.download(file.name)
```

**Comment by rawmotion1** [comment]
I am trying to delete all my replies to all messages in a specific IM. 

In other words, in my private messages with John, I would like to delete every instance where I replied to one of his messages (or one of my own messages) in a thread. 

Is there a way to do this? 

EDIT: I figured it out

```
from slack_cleaner2 import SlackCleaner
s = SlackCleaner("token", sleep_for=1)

for c in s.ims:
  if c.name == 'john':
    for msg in c.msgs(with_replies=True):
        if msg.user == s.myself:
          msg.delete()
